visitors/go_ast/ast_visitor.go

```diff
@@ -32,6 +31,8 @@ func (t *astVisitor) visit(c Context, m interface{}) Cont {
 		return Continue
 	}
 
+	t.vm.ExtendContext(c, t.vm.Visitor(), m)
```

Comment:
If this mutates the context, do we need to undo the mutation before we return?

Reply:
I have a similar question, from reading the PR it's not clear to me what the new discipline for using a Context is-- any information must be deep-copied out of it before returning?

Reply:
Ah, I see.  When visiting children (with `go_ast.DefaultVisitChildren`), the context is copied by calls to `context.EnterStruct`, `context.EnterArray`, and `context.EnterMapValue`.  So when visiting a node `n`, the workflow is:

1. Mutate the context.
2. Call `EnterNode`.
3. Call `VisitChildren`, which copies the context and extends the copy before visiting each child.
4. Call `LeaveNode`.

This ensures that all the visitor methods see the updated context.  We don't need to undo the mutation because  either (a) the context is a copy and will be discarded before visiting any siblings, or (b) there are no siblings, in which case the context is no longer used.

It's equivalent to pushing/popping the context, except the push/pop happens immediately prior to visiting each child rather than on entry to a node.

Reply:
@liujed is there an example somewhere of overriding one of the `Visit*Children` methods for an HTTP/REST visitor?  Is it possible to override one of those methods and forget to copy the context before visiting children?

Reply:
Here's an example of overriding `Visit*Children` for the pair visitor. Currently, if you override `Visit*Children`, before visiting each child, you need to remember to create a context for that child. I suppose we could roll the two operations into `VisitStructField`, `VisitArrayElement`, and `VisitMapValue` methods that update the context for you.

Reply:
What about having the visitor infrastructure copy the context before calling `Visit*Children`, and changing `context.EnterArray` etc. to mutate rather than copy?  That should result in roughly the same number of copy operations, while preventing users of the infrastructure from making a subtle mistake in forgetting to copy the context.

Reply:
Not sure that's enough. Implementations of `Visit*Children` still need to make sure they visit each child with a fresh context, or else the context state will be tainted across sibling nodes.